File: typescript/luhn/luhn.ts
export default class Luhn {
    public static valid(input: string) {
        const addends = this.getAddends(input.replace(/ /g, ''))
        return addends.length > 1 && this.calculateChecksum(addends) % 10 === 0
    }

    private static getAddends(input: string) {
        const addends = []
        let digits = Array.from(input, Number)

        for (let i = 0; i < digits.length; i++)
            addends.push(this.getDigitValue(digits[digits.length - i - 1], i))

        return addends.reverse()
    }

    private static getDigitValue(digit: number, index: number) {
        if (index % 2 === 0) {
            return digit
        }

        const doubledValue = digit * 2
        return doubledValue > 9 ? doubledValue - 9 : doubledValue
    }

    private static calculateChecksum(addends: number[]) {
        return addends.reduce((x, y) => x + y, 0)
    }
}

File: clojure/luhn/src/luhn.clj
(ns luhn (:require [clojure.string :as str]))

(defn valid? [input]
  (let [digits (map #(Character/digit^char % 10) (str/replace input " " ""))
        multipliers (cycle (if (even? (count digits)) [2 1] [1 2]))
        doubled-digits (map * digits multipliers)
        normalized-digits (map #(if (> % 9) (- % 9) %) doubled-digits)
        sum (reduce + normalized-digits)]
    (and (> (count digits) 1) (zero? (mod sum 10)))))
